server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java

```diff
@@ -0,0 +1,183 @@
+         return handler.myStage();
+      }
+      long timeout = (long) (argTimeout * Duration.ofSeconds(1).toMillis());
+
+      // If all the keys are empty or null, create a listener for each key
+      // otherwise return the left value of the first non empty list
+      var pollStage = pollAllKeys(listMultimap, arguments);
+      // If no value returned, we need subscribers
+      return handler.stageToReturn(pollStage.thenCompose(v -> {
+         // addSubscriber call can rise exception that needs to be reported
+         // as error
+         var retStage = (v != null && !v.isEmpty())
+               ? CompletableFuture.completedFuture(v)
+               : addSubscriber(listMultimap, filterKeys, timeout, handler, ctx);
+         return retStage;
+      }), ctx, Consumers.COLLECTION_BULK_BICONSUMER);
+   }
+
+   CompletionStage<Collection<byte[]>> pollKeyValue(EmbeddedMultimapListCache<byte[], byte[]> mmList, byte[] key) {
+      return mmList.pollFirst(key, 1)
+            .thenApply((v) -> (v == null || v.isEmpty())
+                  ? null
+                  : Arrays.asList(key, v.iterator().next()));
+   }
+
+   CompletionStage<Collection<byte[]>> addSubscriber(EmbeddedMultimapListCache<byte[], byte[]> listMultimap,
+         List<byte[]> filterKeys, long timeout, Resp3Handler handler,
+         ChannelHandlerContext ctx) {
+      if (log.isTraceEnabled()) {
+         log.tracef("Subscriber for keys: " +
+               filterKeys.toString());
+      }
+      AdvancedCache<Object, Object> cache = handler.cache().withMediaType(MediaType.APPLICATION_OCTET_STREAM, null);
+      DataConversion vc = cache.getValueDataConversion();
+      PubSubListener pubSubListener = new PubSubListener(ctx.channel());
+      EventListenerKeysFilter filter = new EventListenerKeysFilter(filterKeys.toArray(byte[][]::new));
+      CacheEventConverter<Object, Object, Object> converter = (key, oldValue, oldMetadata, newValue, newMetadata,
+            eventType) -> vc.fromStorage(newValue);
+      CompletionStage<Void> addListenerStage = cache.addListenerAsync(pubSubListener, filter, converter);
+      return CompletionStages.handleAndCompose(addListenerStage, (ignore, t) -> {
+         if (t != null) {
+            pubSubListener.getFuture().completeExceptionally(t);
+         }
+         // Listener can lose events during its install, so we need to poll again
+         // and if we get values complete the listener future.
+         return pollAllKeys(listMultimap, filterKeys).thenCompose(v -> {
+            if (v != null) {
+               // poll return values, complete the listener future ...
+               pubSubListener.getFuture().complete(v);
+            } else {
+               // ... else wait for values with timeout if required
+               pubSubListener.startTimer(timeout);
+               pubSubListener.getFuture().thenApply(ret -> {
+                  pubSubListener.deleteTimer();
+                  return ret;
+               });
```

Comment:
We want `.whenComplete` here. Otherwise, if the listener future completes exceptionally, the lambda never executes.

But since we only ever complete exceptionally if `t != null`, we can skip all this timeout install.

Reply:
I reworked a little bit and added a test to check failure in OnEvent